### PR TITLE
UIQM-331 - Edit MARC authority: Handling updates 1XX or 010 $a value 

### PR DESCRIPTION
**Pull request by Terala-Priyanka**
## Purpose
Edit MARC authority: Handling updates 1XX or 010 $a value
Issue: https://issues.folio.org/browse/UIQM-331

#### TODOS and Open Questions
All the scenarios explained in the scope of the ticket are working fine.
While testing the implementation from various perspectives, I came across an issue in case of a scenario where update 010 or/and 1xx and a delete on some other field happens and then one of the save buttons is clicked.

This have been discussed with Khalilah.
Two approaches have been proposed in order to resolve this issue. The decision will be made by Khalilah on this, and it will be handled in the scope of another ticket.

## Recording
nqEVHaDirD

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.
- [ ] I've added appropriate record to the CHANGELOG.md
- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] If any API-related changes - okapi interfaces and permissions are reviewed/changed correspondingly
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Review by JohnC-80 — APPROVED** [review]

File: src/QuickMarcEditor/QuickMarcEditor.js [hunk 1]
```diff
@@ -135,7 +136,20 @@ const QuickMarcEditor = ({
...
+    continueAfterSave.current = isKeepEditing;
+
+    if (marcType === MARC_TYPES.AUTHORITY && numOfLinks > 0 && are010Or1xxUpdated(initialValues.records, records)) {
+      setIsUpdate0101xxfieldsAuthRecModalOpen(true);
+
+      // eslint-disable-next-line no-useless-return
+      return;
+    } else {
+      confirmSubmit(e);
+    }
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/QuickMarcEditor.js:151]: You do not need `return` here if you use `else`
```suggestion
    if (marcType === MARC_TYPES.AUTHORITY && numOfLinks > 0 && are010Or1xxUpdated(initialValues.records, records)) {
      setIsUpdate0101xxfieldsAuthRecModalOpen(true);
    } else {
      confirmSubmit(e);
    }
```

File: src/QuickMarcEditor/QuickMarcEditor.js [hunk 2]
```diff
@@ -135,7 +136,20 @@ const QuickMarcEditor = ({
     handleSubmit(e).then((updatedRecord) => {
       handleSubmitResponse(updatedRecord);
     });
-  }, [deletedRecords, handleSubmit, handleSubmitResponse]);
+  }, [deletedRecords.length, handleSubmit, handleSubmitResponse]);
+
+  const handleFooterSave = useCallback((e, isKeepEditing = false) => {
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/QuickMarcEditor.js:141]: What does `footer save` mean? Is it a form saving? If yes then maybe `onFormSubmit` will be more readable

File: src/QuickMarcEditor/QuickMarcEditor.test.js [hunk 3]
```diff
@@ -322,6 +325,594 @@ describe('Given QuickMarcEditor', () => {
...
+                  _isLinked: false,
+                  content: '$a Yuan, Bing',
+                  id: 'e95af4e5-008c-42c4-999c-4e103da9de13',
+                  indicators: ['1', '\\'],
+                  isProtected: false,
+                  tag: '100',
+                },
+              ],
+            },
+          });
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/QuickMarcEditor.test.js:405]: It duplicates lines 332-363, I suggest moving this to `quickMarcEditorProps` constant and reuse

File: src/QuickMarcEditor/QuickMarcEditor.test.js [hunk 4]
```diff
@@ -322,6 +325,594 @@ describe('Given QuickMarcEditor', () => {
...
+          fireEvent.change(contentField, { target: { value: '1xx update' } });
+          fireEvent.click(getByText('ui-quick-marc.record.save.continue'));
+          fireEvent.click(getByText('Confirm'));
+
+          expect(onSubmitMock).toHaveBeenCalled();
+          waitFor(() => expect(queryByText('Confirmation modal')).toBeNull());
+          expect(getByText('ui-quick-marc.authority-record.edit.title')).toBeDefined();
+        });
+
+        it('should close the modal on clickinh keep editing button ', () => {
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/QuickMarcEditor.test.js:550]:
```suggestion
        it('should close the modal on clicking keep editing button ', () => {
```

File: src/QuickMarcEditor/QuickMarcEditor.test.js [hunk 5]
```diff
@@ -322,6 +325,594 @@ describe('Given QuickMarcEditor', () => {
...
+
+          const contentField = getByTestId('content-field-1');
+
+          fireEvent.change(contentField, { target: { value: '1xx update' } });
+          fireEvent.click(getByText('stripes-acq-components.FormFooter.save'));
+
+          expect(getByText('Confirmation modal')).toBeDefined();
+        });
+
+        it('should close the modal, save the updates and close the editor on clickng save and close button', async () => {
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/QuickMarcEditor.test.js:630]:
```suggestion
        it('should close the modal, save the updates and close the editor on clicking save and close button', async () => {
```

File: src/QuickMarcEditor/QuickMarcEditor.test.js [hunk 6]
```diff
@@ -322,6 +325,594 @@ describe('Given QuickMarcEditor', () => {
...
+
+          const contentField = getByTestId('content-field-1');
+
+          fireEvent.change(contentField, { target: { value: '1xx update' } });
+          fireEvent.click(getByText('stripes-acq-components.FormFooter.save'));
+
+          expect(getByText('Confirmation modal')).toBeDefined();
+        });
+
+        it('should close the modal, save the updates and editor should be open - on clickng save and keep editing button', async () => {
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/QuickMarcEditor.test.js:735]:
```suggestion
        it('should close the modal, save the updates and editor should be open - on clicking save and keep editing button', async () => {
```

File: src/QuickMarcEditor/QuickMarcEditor.test.js [hunk 7]
```diff
@@ -322,6 +325,594 @@ describe('Given QuickMarcEditor', () => {
...
+          fireEvent.change(contentField, { target: { value: '1xx update' } });
+          fireEvent.click(getByText('ui-quick-marc.record.save.continue'));
+          fireEvent.click(getByText('Confirm'));
+
+          expect(onSubmitMock).toHaveBeenCalled();
+          waitFor(() => expect(queryByText('Confirmation modal')).toBeNull());
+          expect(getByText('ui-quick-marc.authority-record.edit.title')).toBeDefined();
+        });
+
+        it('should close the modal on clickinh keep editing button ', () => {
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/QuickMarcEditor.test.js:771]:
```suggestion
        it('should close the modal on clicking keep editing button ', () => {
```

File: src/QuickMarcEditor/utils.js [hunk 8]
```diff
@@ -807,3 +807,37 @@ export const splitFields = marcRecord => {
     }),
   };
 };
+
+export const are010Or1xxUpdated = (initial, updated) => {
+  let is010Updated = false;
+  let is1XXUpdated = false;
+
+  const initial010 = initial.find(rec => rec.tag === '010');
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/utils.js:815]: Move `010` to const, please

File: src/QuickMarcEditor/utils.js [hunk 9]
```diff
@@ -807,3 +807,37 @@ export const splitFields = marcRecord => {
...
+  const updated010 = updated.find(rec => rec.tag === '010');
+
+  if (initial010 &&
+    updated010 &&
+    getContentSubfieldValue(initial010.content).$a !== getContentSubfieldValue(updated010.content).$a
+  ) {
+    is010Updated = true;
+  }
+
+  const initial1xxRecords = initial.filter(rec => rec.tag[0] === '1');
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/utils.js:825]: Move `1` to const, please

File: src/QuickMarcEditor/utils.js [hunk 10]
```diff
@@ -807,3 +807,37 @@ export const splitFields = marcRecord => {
...
+  const updated1xxRecords = updated.filter(rec => rec.tag[0] === '1');
+
+  const updated1xxArr = [];
+
+  initial1xxRecords.forEach(recI => {
+    const updatedRec = updated1xxRecords.filter(recU => recU.id === recI.id);
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/utils.js:831]: What do `recI` and `recU` mean?

Comment (mariia-aloshyna) [inline src/QuickMarcEditor/utils.js:831]: If you use `.filter` then you expect a new array with filtered records, I guess in this case the constant should be named as `updatedRecords`. If you expect 1 record then use `.find` instead.

File: src/QuickMarcEditor/utils.test.js [hunk 11]
```diff
@@ -1462,4 +1462,64 @@ describe('QuickMarcEditor utils', () => {
...
+      const updated = [
+        {
+          content: 'n  80161705 ',
+          id: 'c1723942-329c-45d4-ac39-a7ac63e015d4',
+          isProtected: true,
+          tag: '001',
+          _isDeleted: false,
+          _isLinked: false,
+        },
+      ];
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/utils.test.js:1487]:
```suggestion
      const recordContent = {
          content: 'n  80161705 ',
          id: 'c1723942-329c-45d4-ac39-a7ac63e015d4',
          isProtected: true,
          tag: '001',
          _isDeleted: false,
          _isLinked: false,
      };
      const initial = [{...recordContent}];
      const updated = [{...recordContent}];
```

File: src/QuickMarcEditor/utils.test.js [hunk 12]
```diff
@@ -1462,4 +1462,64 @@ describe('QuickMarcEditor utils', () => {
...
+          'indicators': [
+            '1',
+            '\\',
+          ],
+          'isProtected': false,
+          'id': '141fc876-f0eb-40ac-9385-38b9de9a3aeb',
+          '_isDeleted': false,
+          '_isLinked': false,
+        },
+      ];
```
Comment (mariia-aloshyna) [inline src/QuickMarcEditor/utils.test.js:1520]:
```suggestion
      const recordContent = {
          tag: '100',
          content: '$a Kitao, Masayoshi, $d 1764-18245',
          indicators: [
            '1',
            '\\',
          ],
          isProtected: false,
          id: '141fc876-f0eb-40ac-9385-38b9de9a3aeb',
          _isDeleted: false,
          _isLinked: false,
      };
      const initial = [{...recordContent}];
      const updated = [{
          ...recordContent,
          content: '$a Kitao, Masayoshi, $d 1764-1824',
      }];
```

File: src/QuickMarcEditor/utils.js (hunk 10 above)
Comment (Terala-Priyanka) [reply src/QuickMarcEditor/utils.js:831]: recI indicates a record object in initial1xxRecords
recU indicates a record object in updated1xxRecords

File: src/QuickMarcEditor/QuickMarcEditWrapper.js [hunk 13]
```diff
@@ -132,7 +153,41 @@ const QuickMarcEditWrapper = ({
 
         setHttpError(parsedError);
       });
-  }, [showCallout, refreshPageData, location, initialValues, instance, locations, marcType, mutator]);
+  }, [showCallout, refreshPageData, location, initialValues, instance, locations, marcType, mutator, numOfLinks]);
+
+  useEffect(() => {
```
Comment (BogdanDenis) [inline src/QuickMarcEditor/QuickMarcEditWrapper.js:158]: Hey Priyanka, does the snippet from this comment have any issues?
https://github.com/folio-org/ui-quick-marc/pull/417#discussion_r1044385960

File: src/QuickMarcEditor/QuickMarcEditWrapper.test.js [hunk 14]
```diff
@@ -431,6 +448,19 @@ describe('Given QuickMarcEditWrapper', () => {
...
+      const mockOnClose = jest.fn();
+
+      renderQuickMarcEditWrapper({
+        instance,
+        mutator,
+        marcType: MARC_TYPES.AUTHORITY,
+        onClose: mockOnClose,
+      });
+
+      expect(mutator.quickMarcInstanceLinks.POST).toHaveBeenCalled();
```
Comment (BogdanDenis) [inline src/QuickMarcEditor/QuickMarcEditWrapper.test.js:461]: Please also check it was called with correct data

File: src/QuickMarcEditor/QuickMarcEditor.js [hunk 15]
```diff
@@ -398,6 +416,25 @@ const QuickMarcEditor = ({
...
         )
       }
+      <ConfirmationModal
+        id="quick-marc-update-linked-bib-fields"
+        open={isUpdate0101xxfieldsAuthRecModalOpen}
+        heading={<FormattedMessage id="ui-quick-marc.update-linked-bib-fields.modal.label" />}
+        message={
+          <FormattedMessage
+            id={continueAfterSave.current ? 'ui-quick-marc.update-linked-bib-fields.modal.message-save-and-editing' : 'ui-quick-marc.update-linked-bib-fields.modal.message-save-and-close'}
+            values={{ count: numOfLinks, saveName: continueAfterSave.current ? 'Save & editing' : 'Save & close' }}
```
Comment (BogdanDenis) [inline src/QuickMarcEditor/QuickMarcEditor.js:426]: Please remove `saveName` from values

File: src/QuickMarcEditor/utils.js (hunk 8 above)
Comment (BogdanDenis) [reply src/QuickMarcEditor/utils.js:815]: Constant can be called `CONTROL_NUMBER_FIELD`

**Review by BogdanDenis — CHANGES_REQUESTED** [review]

File: src/QuickMarcEditor/QuickMarcEditWrapper.js (hunk 13 above)
Comment (Terala-Priyanka) [reply src/QuickMarcEditor/QuickMarcEditWrapper.js:158]: A callout was coming up on page load. Not sure of the reason.

File: src/QuickMarcEditor/QuickMarcEditor.js (hunk 15 above)
Comment (Terala-Priyanka) [reply src/QuickMarcEditor/QuickMarcEditor.js:426]: done.

**Comment by Terala-Priyanka** [comment]
@BogdanDenis and @mariia-aloshyna 
I have addressed most of your comments. For others, I have provided my observations/opinion.

**Review by mariia-aloshyna — APPROVED** [review]
LGTM

**Review by BogdanDenis — APPROVED** [review]

File: src/QuickMarcEditor/utils.js [hunk 16]
```diff
@@ -807,3 +807,39 @@ export const splitFields = marcRecord => {
...
+  if (initial010 &&
+    updated010 &&
+    getContentSubfieldValue(initial010.content).$a !== getContentSubfieldValue(updated010.content).$a
+  ) {
+    is010Updated = true;
+  }
+
+  const authRec1XXTagStartsWith = '1';
+  const initial1xxRecords = initial.filter(rec => rec.tag[0] === authRec1XXTagStartsWith);
+  const updated1xxRecords = updated.filter(rec => rec.tag[0] === authRec1XXTagStartsWith);
```
Comment (Dmytro-Melnyshyn) [inline src/QuickMarcEditor/utils.js:828]: Please use `find`, an authority record can't have multiple 1xx fields.

File: src/QuickMarcEditor/QuickMarcEditWrapper.js [hunk 17]
```diff
@@ -132,7 +153,22 @@ const QuickMarcEditWrapper = ({
...
+    if (marcType !== MARC_TYPES.AUTHORITY) {
+      return;
+    }
+
+    mutator.quickMarcInstanceLinks.POST({
```
Comment (Dmytro-Melnyshyn) [inline src/QuickMarcEditor/QuickMarcEditWrapper.js:163]: It should be invoked only for an authorized record.

Comment (Dmytro-Melnyshyn) [reply src/QuickMarcEditor/QuickMarcEditWrapper.js:163]: const authRefType = new URLSearchParams(location.search).get('authRefType');

    if (marcType === MARC_TYPES.AUTHORITY && authRefType === 'Authorized')

File: src/QuickMarcEditor/QuickMarcEditWrapper.js [hunk 18]
```diff
@@ -132,7 +153,22 @@ const QuickMarcEditWrapper = ({
...
+      'ids': [instance.id],
+    })
+      .catch(async e => {
+        const parsedError = await parseHttpError(e);
```
Comment (Dmytro-Melnyshyn) [inline src/QuickMarcEditor/QuickMarcEditWrapper.js:167]: `parseHttpError` can't handle `content-type: application/json`

**Pull request merged**